src/generic_factory.h

```diff
@@ -223,25 +296,63 @@ class generic_factory
         /**
          * Returns the object with the given id.
          * The input id should be valid, otherwise a debug message is issued.
+         * This function can be used to implement @ref int_id::obj().
+         * Note: If the id was valid, the returned object can be modified (after
+         * casting the const away).
+         */
+        const T &obj( const int_id<T> &id ) const {
+            if( !is_valid( id ) ) {
+                debugmsg( "invalid %s id \"%d\"", type_name.c_str(), id );
+                return dummy_obj();
+            }
+            return list[id];
+        }
+        /**
+         * Returns the object with the given id.
+         * The input id should be valid, otherwise a debug message is issued.
          * This function can be used to implement @ref string_id::obj().
          * Note: If the id was valid, the returned object can be modified (after
          * casting the const away).
          */
-        const T &obj( const string_id<T> &sid ) const {
-            const auto iter = data.find( sid );
-            if( iter == data.end() ) {
-                debugmsg( "invalid %s id \"%s\"", type_name.c_str(), sid.c_str() );
-                static const T dummy{};
-                return dummy;
+        const T &obj( const string_id<T> &id ) const {
+            static int_id<T> i_id;
```

Comment:
But it's confusing and error prone. Constructing (and initializing) a single `int` (which is all that `int_id` contains) shouldn't take any time.

I actually suspect this will prevent the compiler from certain optimizations: a non-static variable (especially since this is effectively only an `int) can be put into a register. The static variable must be in main memory and may need to be loaded again and again.

Also: the compiler may see that id is immediately (and unconditionally) overwritten in `find` and may skip the initialization.

In short: looks like premature optimization.